resources/testdriver.js

```diff
@@ -0,0 +1,101 @@
+        var right = Math.min(window.innerWidth, rect.right);
+        var top = Math.max(0, rect.top);
+        var bottom = Math.min(window.innerHeight, rect.bottom);
+
+        var x = 0.5 * (left + right);
+        var y = 0.5 * (top + bottom);
+
+        return [x, y];
+    }
+
+    function getPointerInteractablePaintTree(element) {
+        if (!window.document.contains(element)) {
+            return [];
+        }
+
+        var rectangles = element.getClientRects();
+
+        if (rectangles.length === 0) {
+            return [];
+        }
+
+        var centerPoint = getInViewCenterPoint(rectangles[0]);
+
+        if ("elementsFromPoint" in document) {
+            return document.elementsFromPoint(centerPoint[0], centerPoint[1]);
+        } else {
+            var rv = document.msElementsFromPoint(centerPoint[0], centerPoint[1]);
+            return Array.prototype.slice.call(rv ? rv : []);
+        }
+    }
+
+    function inView(element) {
+        var pointerInteractablePaintTree = getPointerInteractablePaintTree(element);
+        return pointerInteractablePaintTree.indexOf(element) !== -1;
+    }
+
+
+    /**
+     * @namespace
+     */
+    window.test_driver = {
+        /**
+         * Triggers a user-initiated click
+         *
+         * This matches the behaviour of the {@link
+         * https://w3c.github.io/webdriver/webdriver-spec.html#element-click|WebDriver
+         * Element Click command}.
+         *
+         * @param {Element} element - element to be clicked
+         * @returns {Promise} fulfilled after click occurs, or rejected in
+         *                    the cases the WebDriver command errors
+         */
+        click: function(element) {
+            if (window.top !== window) {
+                return Promise.reject(new Error("can only click in top-level window"));
+            }
+
+            if (!window.document.contains(element)) {
+                return Promise.reject(new Error("element in different document or shadow tree"));
+            }
+
+            if (!inView(element)) {
+                element.scrollIntoView({behavior: "instant",
+                                        block: "end",
+                                        inline: "nearest"});
+            }
+
+            var pointerInteractablePaintTree = getPointerInteractablePaintTree(element);
+            if (pointerInteractablePaintTree.length === 0 ||
+                !element.contains(pointerInteractablePaintTree[0])) {
+                return Promise.reject(new Error("element click intercepted error"));
+            }
+
+            var rect = element.getClientRects()[0];
+            var top = Math.min(rect.y, rect.y + rect.height);
+            var left = Math.min(rect.x, rect.x + rect.width);
+            var centerPoint = getInViewCenterPoint(rect);
+            return window.test_driver_internal.click(element,
+                                                     {x: top + centerPoint[0],
```

Comment:
I was playing with this code to get it to work on Chrome. Do you think this top and left in the next line should be removed? It is already calculated inside getInViewCenterPoint function?

Reply:
We probably want to do something nicer than this, I would agree.

Reply:
I tried to fix it [here](https://chromium-review.googlesource.com/c/chromium/src/+/754063) already. Let me know if you think I mis-understood this part.

Reply:
Oh, you mean like that. Then no, because the coordinates passed as arguments here are relative to the viewport, whereas centerPoint are relative to the top-left corner of the element. That's a deliberate difference (because it means you don't need to find the top-left corner of the element again).

Reply:
I'm not sure I'm following. Say your viewport is whole window and there is no zoom or anything and the element is also inside the viewport. If you follow the math here you see that top is being added twice here which makes it off the element.

Reply:
I think I'm just being stupid here and you're right, but it's late so *shrugs and delays until next week*.

Reply:
https://github.com/w3c/web-platform-tests/pull/8057 is likely going to be automatically merged soon. Please continue there if needed :)